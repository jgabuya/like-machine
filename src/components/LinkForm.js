import React from "react";
import { Form, FormGroup, Label, Input, Button } from "reactstrap";
import PropTypes from "prop-types";

const propTypes = {	
	onUrlChange: PropTypes.func.isRequired,
	onSubmit: PropTypes.func.isRequired
};

const LinkForm = props => {
	return (
		<div>
			<Form onSubmit={props.onSubmit}>
				<h4>New Link</h4>

				<FormGroup>
					<Label for="input-title">URL</Label>
					<Input
						type="text"
						id="input-title"
                        onChange={props.onUrlChange}
                        required
					/>
				</FormGroup>

				<FormGroup>
					<Button color="primary">Submit</Button>
				</FormGroup>
			</Form>
		</div>
	);
};

LinkForm.propTypes = propTypes;

export default LinkForm;
