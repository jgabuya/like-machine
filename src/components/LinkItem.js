import React, { Component } from "react";
import PropTypes from "prop-types";
import {
    Card,
    CardBody,
    CardTitle,
    CardSubtitle,
    CardText,
    CardFooter,
    Button,
    CardImg
} from "reactstrap";
import { FaHeart, FaRegHeart, FaRegTrashAlt } from "react-icons/fa";

import moment from "moment";
import normalizeUrl from "normalize-url";

import "animate.css";

const propTypes = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
    description: PropTypes.string,
    url: PropTypes.string.isRequired,
    imageUrl: PropTypes.string,
    createdAt: PropTypes.string.isRequired,
    liked: PropTypes.bool.isRequired,
    likeCount: PropTypes.number.isRequired,
    owned: PropTypes.bool.isRequired,
    onLike: PropTypes.func.isRequired,
    onUnlike: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired
};

class LinkItem extends Component {
    state = {
        animatedLike: false
    };

    componentWillReceiveProps(nextProps) {
        this.setState({
            animatedLike: nextProps.liked && !this.props.liked
        });
    }

    render() {
        let LikeIcon;
        let likeBtnTitle;
        let likeBtnOnClickHandler;

        if (this.props.liked) {
            LikeIcon = FaHeart;
            likeBtnTitle = "Unlike";
            likeBtnOnClickHandler = this.props.onUnlike;
        } else {
            LikeIcon = FaRegHeart;
            likeBtnTitle = "Like";
            likeBtnOnClickHandler = this.props.onLike;
        }

        return (
            <Card className="mb-5">
                {this.props.imageUrl !== null && (
                    <CardImg
                        top
                        src={this.props.imageUrl}
                        alt={this.props.title || this.props.url}
                        style={{
                            width: "100%",
                            maxHeight: "20vw",
                            objectFit: "cover"
                        }}
                    />
                )}

                <CardBody>
                    <CardTitle>
                        <a target="_blank" href={normalizeUrl(this.props.url)}>
                            {this.props.title || this.props.url}
                        </a>
                    </CardTitle>

                    <CardSubtitle>{this.props.description}</CardSubtitle>

                    <CardText>
                        <small
                            className="text-muted"
                            title={this.props.createdAt}
                        >
                            {moment(this.props.createdAt)
                                .startOf("day")
                                .fromNow()}
                        </small>
                    </CardText>
                </CardBody>

                {this.props.user && (
                    <CardFooter>
                        <Button
                            onClick={likeBtnOnClickHandler}
                            outline
                            color="primary"
                            title={likeBtnTitle}
                        >
                            <LikeIcon
                                className={
                                    this.state.animatedLike
                                        ? "animated bounce"
                                        : ""
                                }
                            />{" "}
                            <small>{this.props.likeCount}</small>
                        </Button>

                        {this.props.owned && (
                            <Button
                                onClick={this.props.onDelete}
                                outline
                                color="danger"
                                className="float-right"
                                title="Delete"
                            >
                                <FaRegTrashAlt />
                            </Button>
                        )}
                    </CardFooter>
                )}
            </Card>
        );
    }
}

LinkItem.propTypes = propTypes;

export default LinkItem;
