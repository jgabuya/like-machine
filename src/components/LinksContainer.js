import React, { Component } from 'react';
import Link from './Link/Link';

import axios from '../util/axios-wrapper'

class LinksContainer extends Component {
    constructor(props) {
        super(props)

        this.state = {
            links: []
        }
    }

    async componentDidMount() {
        const result = await axios.get('/links');

        this.setState({
            links: result.data
        });
    }

    renderLinks() {
        return this.state.links.map((item, index) => {
            return (
                <Link
                    key={ index }
                    id={ item.id }
                    title={ item.title }
                    description={ item.description }
                    url={ item.url }
                    createdAt={ item.created_at }
                    liked={ item.liked }
                    likeCount={ item.like_count }
                    owned={ item.owned }
                    onLike={ () => {} }
                    onDelete={ () => {} }
                />
            )
        })
    }
    
    render() {
        return (
            <div>
                { this.renderLinks() }
            </div>
        );
    }
}

export default LinksContainer;
