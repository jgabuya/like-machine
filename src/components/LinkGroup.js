import React from "react";
import PropTypes from "prop-types";
import LinkItem from "./LinkItem";

import { sortLinksByLikeCountDesc } from "../util/data-operations";
import moment from "moment";
import { confirmAlert } from "react-confirm-alert";
import "react-confirm-alert/src/react-confirm-alert.css";

const propTypes = {
    linkGroup: PropTypes.object.isRequired,
    onDelete: PropTypes.func.isRequired,
    onLike: PropTypes.func.isRequired,
    onUnlike: PropTypes.func.isRequired
};

const renderLinks = (links, onLike, onUnlike, onDelete) => {
    return links.map((item, index) => {
        return (
            <LinkItem
                key={index}
                id={item.id}
                title={item.title}
                description={item.description}
                url={item.url}
                imageUrl={item.image_url}
                createdAt={item.created_at}
                liked={item.liked}
                likeCount={item.like_count}
                owned={item.owned}
                onLike={() => {
                    onLike(item.id);
                }}
                onUnlike={() => {
                    onUnlike(item.id);
                }}
                onDelete={() => {
                    confirmAlert({
                        title: "Confirm delete",
                        message: "Are you sure?",
                        buttons: [
                            {
                                label: "Yes",
                                onClick: () => onDelete(item.id)
                            },
                            {
                                label: "No"
                            }
                        ]
                    });
                }}
            />
        );
    });
};

const LinkGroup = props => {
    return (
        <div>
            <h4 className="mb-3 text-muted">
                {moment.unix(props.linkGroup.timestamp).format("MMMM D, YYYY")}
            </h4>

            {renderLinks(
                sortLinksByLikeCountDesc(props.linkGroup.links),
                props.onLike,
                props.onUnlike,
                props.onDelete
            )}
        </div>
    );
};

LinkGroup.propTypes = propTypes;

export default LinkGroup;
