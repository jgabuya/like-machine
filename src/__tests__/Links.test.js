import React from 'react';
import { Links } from '../components';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('Links', () => {
    const props = {
        links: []
    };

    it('matchesSnapshot', () => {
        const renderedValue = renderer.create(<Links {...props} />).toJSON();
        expect(renderedValue).toMatchSnapshot();
    })

    it('renders without crashing', () => {
        const wrapper = shallow(<Links {...props} />);
        expect(wrapper.length).toEqual(1);
    });    
})
