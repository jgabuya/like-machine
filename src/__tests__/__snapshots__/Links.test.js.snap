// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links matchesSnapshot 1`] = `<div />`;
